File: src/main/java/io/github/jho951/assignment/job/processing/JobProcessingScheduler.java
package io.github.jho951.assignment.job.processing;

import java.util.concurrent.Executor;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;

@Component
@ConditionalOnProperty(name = "jobs.scheduling-enabled", havingValue = "true", matchIfMissing = true)
@RequiredArgsConstructor
public class JobProcessingScheduler {

    private final JobProcessor jobProcessor;
    private final JobRecoveryService jobRecoveryService;
    private final Executor jobTaskExecutor;

    @Scheduled(fixedDelayString = "${jobs.poll-interval-ms}")
    public void processDueJobs() {
        jobRecoveryService.recoverStaleJobs();

        for (String jobId : jobProcessor.findDueJobIds()) {
            if (jobProcessor.claimJobForProcessing(jobId)) {
                jobTaskExecutor.execute(() -> jobProcessor.processClaimedJob(jobId));
            }
        }
    }
}

File: src/main/kotlin/io/github/jho951/assignment/job/processing/JobProcessingScheduler.kt
package io.github.jho951.assignment.job.processing

import java.util.concurrent.Executor
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

@Component
@ConditionalOnProperty(name = ["jobs.scheduling-enabled"], havingValue = "true", matchIfMissing = true)
class JobProcessingScheduler(
    private val jobProcessor: JobProcessor,
    private val jobRecoveryService: JobRecoveryService,
    private val jobTaskExecutor: Executor
) {

    @Scheduled(fixedDelayString = "\${jobs.poll-interval-ms}")
    fun processDueJobs() {
        jobRecoveryService.recoverStaleJobs()

        for (jobId in jobProcessor.findDueJobIds()) {
            if (jobProcessor.claimJobForProcessing(jobId)) {
                jobTaskExecutor.execute { jobProcessor.processClaimedJob(jobId) }
            }
        }
    }
}
